exit/test-exit.c

```diff
@@ -0,0 +1,1071 @@
+	/* child */
+	if (pid == 0) {
+		pthread_t tid;
+
+		close(pipefd[0]);
+
+		ret = pthread_create(&tid, NULL, test_threadFunc, ptr);
+		write(pipefd[1], &ret, sizeof(int));
+		close(pipefd[1]);
+		free(ptr);
+		test_common.test_threadWait = 1;
+		pthread_join(tid, NULL);
+	}
+	/* parent */
+	else {
+		int status;
+
+		ret = wait(&status);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+
+		close(pipefd[1]);
+		/* Get return value from pthread_create() */
+		TEST_ASSERT_EQUAL_INT(sizeof(int), read(pipefd[0], &ret, sizeof(int)));
+		TEST_ASSERT_EQUAL_INT(0, ret);
+
+		TEST_ASSERT_EQUAL_INT(-1, access(TEST_EXIT_PATH, F_OK));
+
+		close(pipefd[0]);
+		free(ptr);
+	}
+}
+
+
+TEST(unistd_exit, no_atexit)
+{
+	/* Test that _exit() do NOT invoke any funcs registered by atexit */
+	pid_t pid;
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		atexit(test_no_atexit);
+		test_common.test_exitPtr(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int status, ret;
+
+		ret = wait(&status);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+
+		TEST_ASSERT_EQUAL_INT(-1, access(TEST_EXIT_PATH, F_OK));
+
+		remove(TEST_EXIT_PATH);
+	}
+}
+
+
+TEST(unistd_exit, no_flush)
+{
+	/* Test that _exit() do NOT force unwritten buffered data to be flushed */
+	pid_t pid;
+	int fd;
+
+	FILE *f = fopen(TEST_EXIT_PATH, "w+");
+	TEST_ASSERT_NOT_NULL(f);
+	fd = open(TEST_EXIT_PATH, O_RDWR);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fd);
+	/* Check file is empty */
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_END));
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		fprintf(f, TEST_EXIT_STR);
+		test_common.test_exitPtr(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int status, ret;
+
+		ret = wait(&status);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+
+		/* If buffered data was flushed file length should increase */
+		TEST_ASSERT_NOT_EQUAL_INT(strlen(TEST_EXIT_STR), lseek(fd, 0, SEEK_END));
+
+		fclose(f);
+		close(fd);
+		remove(TEST_EXIT_PATH);
+	}
+}
+
+
+TEST(unistd_exit, no_handler)
+{
+	/* Test that handler is not invoked when _exit */
+	pid_t pid;
+
+	/* Clear handlerFlag, if handler invoked some dummy value would be set */
+	test_common.test_handlerFlag = 0;
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		struct sigaction sa;
+
+		sa.sa_handler = test_sigusrHandler;
+		TEST_ASSERT_EQUAL_INT(0, sigemptyset(&sa.sa_mask));
+		sa.sa_flags = 0;
+		TEST_ASSERT_EQUAL_INT(0, sigaction(SIGUSR1, &sa, NULL));
+
+		test_common.test_exitPtr(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int status, ret;
+
+		ret = wait(&status);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+
+		TEST_ASSERT_EQUAL_INT(0, test_common.test_handlerFlag);
+	}
+}
+
+
+TEST_GROUP(stdlib_exit);
```

Comment:
From `exit` doc:

If a function registered by a call to atexit() fails to return, the remaining registered functions shall not be called and the rest of the exit() processing shall not be completed

IMHO it may also be tested.

Reply:
IMHO it is pointless.

Reply:
Ok, it would be hard.